apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PooledSession.cs

```diff
@@ -315,17 +556,49 @@ public void DetachFromPool()
         /// from this object.</param>
         /// <param name="callSettings">If not null, applies overrides to this RPC call.</param>
         /// <returns>A task representing the asynchronous operation. When the task completes, the result is the response from the RPC.</returns>
-        public async Task<CommitResponse> CommitAsync(CommitRequest request, CallSettings callSettings)
+        public Task<CommitResponse> CommitAsync(CommitRequest request, CallSettings callSettings)
         {
             CheckNotDisposed();
             GaxPreconditions.CheckNotNull(request, nameof(request));
-            GaxPreconditions.CheckState(TransactionId != null, "Cannot commit a PooledSession with no associated transaction");
+
             request.SessionAsSessionName = SessionName;
-            request.TransactionId = TransactionId;
 
-            var response = await RecordSuccessAndExpiredSessions(Client.CommitAsync(request, callSettings)).ConfigureAwait(false);
-            MarkAsCommittedOrRolledBack();
-            return response;
+            return ExecuteMaybeWithTransactionSelectorAsync(
+                transactionSelectorSetter: SetTransaction,
+                commandAsync: CommitAsync,
+                inlinedTransactionExtractor: null, // Commit does not support inline transactions.
+                skipTransactionCreation: request.Mutations.Count == 0, // If there are only mutations we won't have a transaction but we need one.
+                callSettings?.CancellationToken ?? default);
+
+            void SetTransaction(TransactionSelector transactionSelector) 
```

Comment:
Maybe rename this (and similar methods) to SetTransactionInRequest or similar, so that we know "SetTransaction" always means "set transaction in the pooled session".